pyscriptjs/src/plugins/pyterminal.ts

```diff
@@ -70,59 +83,182 @@ export class PyTerminalPlugin extends Plugin {
             // implications
             this.outElem = document.createElement('pre');
-            this.outElem.className = 'py-terminal';
+            this.outElem.classList.add('py-terminal');
             this.appendChild(this.outElem);
 
-            if (this.isAuto()) {
-                this.classList.add('py-terminal-hidden');
-                this.autoShowOnNextLine = true;
-            } else {
-                this.autoShowOnNextLine = false;
-            }
+            this.setupPosition(app);
+        }
 
+        // implementation of the Stdio interface
+        stdout_writeline(msg: string) {
+            this.outElem.innerText += msg + '\n';
             if (this.isDocked()) {
-                this.classList.add('py-terminal-docked');
+                this.scrollTop = this.scrollHeight;
+            }
+            if (this.autoShowOnNextLine) {
+                this.classList.remove('py-terminal-hidden');
+                this.autoShowOnNextLine = false;
             }
+        }
 
-            logger.info('Registering stdio listener');
-            app.registerStdioListener(this);
+        stderr_writeline(msg: string) {
+            this.stdout_writeline(msg);
         }
+        // end of the Stdio interface
+    }
+
+    return PyTerminalPre;
+}
+
+declare const Terminal: typeof TerminalType;
+
+function make_PyTerminal_xterm(app: PyScriptApp) {
+    /** The <py-terminal> custom element, which automatically register a stdio
+     *  listener to capture and display stdout/stderr
+     */
+    class PyTerminalXterm extends PyTerminalBaseClass {
+        outElem: HTMLDivElement;
+        _moduleResolved: boolean;
+        xtermReadyPromise: Promise<TerminalType>;
+        xterm: TerminalType;
+        cachedStdOut: Array<string>;
+        cachedStdErr: Array<string>;
+        _xterm_cdn_base_url = 'https://cdn.jsdelivr.net/npm/xterm@5.1.0';
 
-        isAuto() {
-            return this.hasAttribute('auto');
+        constructor() {
+            super();
+            this.cachedStdOut = [];
+            this.cachedStdErr = [];
+
+            // While this is false, store writes to stdout/stderr to a buffer
+            // when the xterm.js is actually ready, we will "replay" those writes
+            // and set this to true
+            this._moduleResolved = false;
+
+            //Required to make xterm appear properly
+            this.style.width = '100%';
+            this.style.height = '100%';
         }
 
-        isDocked() {
-            return this.hasAttribute('docked');
+        async connectedCallback() {
+            //guard against initializing a tag twice
+            if (knownPyTerminalTags.has(this)) return;
+
+            this.outElem = document.createElement('div');
+            //this.outElem.className = 'py-terminal';
+            this.appendChild(this.outElem);
+
+            this.setupPosition(app);
+
+            this.xtermReadyPromise = this._setupXterm();
```

Comment:
nit: I do prefer non-redundant names ... `connectedCallback` as example is a well known ugly name as that's a callback by definition (you literally define a method for that, and the name is not even `connectedMethod`) ... we also use TS in here and I feel like it's obvious any asynchronous callback/method gonna return a promise ... do we really need to duplicate unnecessarily the name *Promise* for anything asynchronous? we don't in any other place and it would be vary ugly if everything *async* ends up with the `Promise` name ... can we call this property `this.xtermReady` instead where it's obvious it's a Promise either by looking at the code (`await` needed) or via documentation?

Reply:
I don't hate calling it just `this.xtermReady` - will make the change in src and docs.